### ISSUE TITLE
RT-DETR-l TensorRT export (.engine) returns incorrect bounding boxes while .pt and .onnx work (Jetson AGX Thor)

### ISSUE THREAD

**Issue by ashuguptahere** [ultralytics/ultralytics#22873]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Export

### Bug

I’m deploying an RT-DETR-l model trained with Ultralytics on a Jetson AGX Thor using TensorRT 10.13.3.9-1+cuda13.0.

- The PyTorch .pt model works and produces detections.
- The ONNX model works and produces correct bounding boxes.
- The TensorRT .engine model builds successfully, but when I run inference with it in Python, it returns incorrect bounding boxes (attachment attached below)

So effectively: same weights, same image, same preprocessing → .pt and .onnx detect correctly, .engine does not.

### Environment

$ yolo checks
```
Ultralytics 8.3.234 🚀 Python-3.12.0 torch-2.9.1+cu130 CUDA:0 (NVIDIA Thor, 125771MiB)
Setup complete ✅ (14 CPUs, 122.8 GB RAM, 366.3/936.4 GB disk)

OS                     Linux-6.8.12-tegra-aarch64-with-glibc2.39
Environment            Linux
Python                 3.12.0
Install                git
Path                   /home/thor/Desktop/ultralytics-rtdetr-export/.venv/lib/python3.12/site-packages/ultralytics
RAM                    122.82 GB
Disk                   366.3/936.4 GB
CPU                    aarch64
CPU count              14
GPU                    NVIDIA Thor, 125771MiB
GPU count              1
CUDA                   13.0

numpy                  ✅ 1.26.0>=1.23.0
matplotlib             ✅ 3.10.7>=3.3.0
opencv-python          ✅ 4.11.0.86>=4.6.0
pillow                 ✅ 12.0.0>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.16.3>=1.4.1
torch                  ✅ 2.9.1+cu130>=1.8.0
torch                  ✅ 2.9.1+cu130!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.24.1>=0.9.0
psutil                 ✅ 7.1.3>=5.8.0
polars                 ✅ 1.35.2>=0.20.0
ultralytics-thop       ✅ 2.0.18>=2.0.18
```

### Minimal Reproducible Example

```
import sys

from ultralytics import RTDETR

path = f"/usr/lib/python3.12/dist-packages"
sys.path.append(path)

# Load a PyTorch model
model = RTDETR("rtdetr-l.pt")

# Export the model to TensorRT
model.export(format="engine")

# Load the exported TensorRT model
trt_model = RTDETR("rtdetr-l.engine")

# Run inference
results = trt_model("https://ultralytics.com/images/bus.jpg", show=True, save=True)
```

### Additional

I tried different opset settings with different models (rtdetr-l and rtdetr-x models), but it didn't work. I am getting this kind of result after prediction. Also, tried installing different version of torch namely 2.9.0 and 2.9.1, the issue persists

![Image](https://github.com/user-attachments/assets/fec59b24-9cb1-4d1d-a1b7-df9f672589d1)

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ashuguptahere, thank you for your detailed report and for your interest in Ultralytics 🚀! This is an automated response to help get things moving quickly — an Ultralytics engineer will review your issue and assist you soon.

In the meantime, we recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Since this is a 🐛 Bug Report related to export, you’ve already provided a good code snippet 👍. If possible, please also include:
- Exact `yolo export` or Python export command used (with all arguments)
- Any TensorRT build logs or warnings
- A short example (single image and corresponding predictions) comparing `.pt` / `.onnx` vs `.engine` outputs

If you are able to further simplify your script, please ensure it still forms a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it efficiently.

If this turns into a custom training or deployment ❓ Question (for example, if it ends up being configuration-related), please share as much detail as you can, including:
- Model variant and export settings
- Example images and screenshots of outputs
- Any relevant logs, and confirm you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)

You’re also welcome to join the Ultralytics community where it suits you best:
- For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧  
- For in-depth discussions, check out [Discourse](https://community.ultralytics.com/)  
- Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community 💬  

## Upgrade

Before we dig deeper, please upgrade to the latest `ultralytics` package and verify the behavior persists, as many export- and backend-related issues are fixed over time. Make sure all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) are satisfied in a **Python>=3.8** environment with **PyTorch>=1.8**:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

If possible, please also confirm:
- Whether the issue reproduces in one of the verified environments above
- The exact TensorRT build command or toolchain you used for the `.engine` file

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Once you share any additional details or refinements to your minimum reproducible example, an Ultralytics engineer will jump in to help analyze the RT-DETR TensorRT export behavior on your Jetson AGX Thor setup 🔍

**Comment by Y-T-G** [comment]
Does YOLO predictions work with TensorRT on Thor?

**Comment by ashuguptahere** [comment]
Yes, `YOLO` predictions are working fine with `.pt`, `.onnx` and `.engine` too but failing with `RTDETR` models

**Comment by Y-T-G** [comment]
I think you need to ask to NVIDIA forums maybe.

@lakshanthad

**Comment by lakshanthad** [comment]
@ashuguptahere I have verified `RTDETR` models converted to TensorRT on the below hardware:

Consumer GPU: RTX 3060 (CUDA 12.x) -> works
NVIDIA Jetson AGX Orin with JetPack 6 (CUDA 12.x) - works
NVIDIA Jetson Thor with JetPack 7 (CUDA 13.x) - have same issue as yours

Let me debug further on Jetson Thor and get back to you on this.

**Comment by ashuguptahere** [comment]
@lakshanthad any update on the issue?

**Comment by ashuguptahere** [comment]
Furthermore, I tried exporting RTDETR and RTDETRv2 from original repo and also tried RFDETR from roboflow, but the issue seems to be only with RTDETR from ultralytics. Even tried the newest version of ultralytics, but no luck yet.

**Comment by glenn-jocher** [comment]
We can reproduce this on Jetson Thor / JetPack 7 (CUDA 13.x) and are still debugging it (no ETA yet); to help isolate whether it’s a Thor TensorRT quirk vs an environment/import conflict, could you retry with a fixed-shape FP32 engine and share the results, ideally from a clean env like the JetPack 7 container in the [Jetson quickstart guide](https://docs.ultralytics.com/guides/nvidia-jetson/)?

```python
from ultralytics import RTDETR

m = RTDETR("rtdetr-l.pt")
m.export(format="engine", imgsz=640, dynamic=False, half=False, batch=1)  # force FP32 + static shapes

trt = RTDETR("rtdetr-l.engine")
trt.predict("path/to/image.jpg", imgsz=640, verbose=True, save=True)
```

Also please paste the text output of the following (your `sys.path.append(...)` suggests your venv may be mixing with system packages):

```bash
python -c "import tensorrt as trt; print(trt.__version__); print(trt.__file__)"
python -c "from ultralytics.utils.checks import collect_system_info; collect_system_info()"
```

**Comment by lakshanthad** [comment]
@ashuguptahere We are still debugging the issue because the issue is only there for Jetson AGX Thor device running JetPack 7. The issue is not there when using all the other devices we have.

Could you share with me your export code when you said you used original repo of RTDETR? I could try to debug from my side as well to further understand the issue when using Ultralytics to export the model.

We will also try on DGX Spark as well just to check whether it is related to Blackwell architecture GPUs.

**Comment by ashuguptahere** [comment]
Hey @lakshanthad 

I used this code to export RTDETR model from the [original repo](https://github.com/lyuwenyu/RT-DETR/blob/main/rtdetr_pytorch/tools/export_onnx.py) on Jetson Thor device.

From this I was able to export `rtdetr_r50vd_m_6x_coco.yml` and `rtdetr_r50vd_6x_coco.yml` using their config files, but I didn't try the rest.

Also, I had to make changes to `imgsz`, `batchsize` and `opset_version` as well to satisfy my need

I will share the exact code I used after setting up my system again, because sadly I've lost access to all my code which I used for export as I flashed newer JetPack 7.1 thinking it would resolve my problem.

**Comment by lakshanthad** [comment]
@ashuguptahere Thanks for the information. 

@onuralpszr Could you please help to test this on NVIDIA DGX Spark as well? I am not sure this is a problem with CUDA 13 on ARM because all other Jetsons with CUDA 12.x have no issues whereas consumer GPUs with both CUDA 12.x and CUDA 13 have no issues.

**Comment by lakshanthad** [comment]
@ashuguptahere It seems that it is an issue with `tensorrt 10.13.3.9` which comes pre-installed on Jetson AGX Thor devices. The issue is resolved after installing `tensorrt 10.15.1.29`. I have made [a PR](https://github.com/ultralytics/ultralytics/pull/23499) to fix this. We will make this update available for both AGX Thor and DGX Spark devices because both of them are suffering the same problem.

 In the meantime, please do `pip install "tensorrt-cu13>10.15"`

Please check the results below:

![Image](https://github.com/user-attachments/assets/b82e32a6-09dd-49e9-b13e-27f80b7e4b64)
![Image](https://github.com/user-attachments/assets/ed22aa78-90bd-4ae9-8df2-13ce01fdd73e)

cc: @Y-T-G @onuralpszr

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has now been merged in [PR #23499: Force update `tensorrt` on CUDA 13 ARM to `10.15.x` to fix a bug with RT-DETR exports](https://github.com/ultralytics/ultralytics/pull/23499).

### What changed (and why it may resolve this)
This PR improves TensorRT export reliability on NVIDIA Jetson / JetPack 7 (CUDA 13 ARM), with a specific focus on RT-DETR export failures:
- **JetPack 7 detection support**: Updates `is_jetson()` to recognize JetPack 7 by mapping it to **L4T `R38`**, ensuring Jetson environments are detected correctly.
- **More flexible TensorRT version checks**: Updates `check_tensorrt()` to accept a configurable `min_version` (default `7.0.0`), making version gating more adaptable across platforms.
- **Pins to a known-good TensorRT line for JetPack 7 during export**: When exporting TensorRT engines on JetPack 7, it **forces a `10.15.x`-compatible TensorRT requirement**, which is expected to address the RT-DETR export bug seen on these setups.

### Credits
Huge thanks to @lakshanthad for the PR, with contributions from @Laughing-q and @onuralpszr.

### How to test the fix
Since this was just merged, the easiest ways to validate are:

- **Install Ultralytics from `main`** (includes the merged changes):
  ```bash
  pip install -U git+https://github.com/ultralytics/ultralytics.git@main
  ```

- **Or wait for the next official PyPI release** (if preferred), which will include this merge once published.

After updating, please retry the same TensorRT export workflow that was failing (especially RT-DETR on JetPack 7 / CUDA 13 ARM).

### Feedback requested
Confirmation on whether the changes in PR #23499 resolve the issue would be really helpful (including JetPack version, device, CUDA/TensorRT versions, and the exact export command/log snippet if anything still fails).

Thanks 🙏 for reporting this—any further feedback is welcome if the issue persists.
